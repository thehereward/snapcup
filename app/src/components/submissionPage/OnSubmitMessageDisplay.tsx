import React from "react";
import styled from "styled-components";
import { HelperText } from "./SnapSubmissionStyles";

const SnapSuccessText = styled(HelperText)`
    color: var(--text-success);
`;

const OnSubmitMessageDisplay: React.FunctionComponent = (props: {
    confirmation: Boolean;
}) => {
    if (props.confirmation) {
        return <SnapSuccessText>Snap submitted!</SnapSuccessText>;
    } else {
        return <p></p>;
    }
};

export default OnSubmitMessageDisplay;
