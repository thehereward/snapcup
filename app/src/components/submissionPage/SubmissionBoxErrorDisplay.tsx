import React from "react";
import styled from "styled-components";

const ErrorMessage = styled.p`
    text-align: right;
    font-family: Open Sans;
    font-style: normal;
    font-weight: 600;
    font-size: 12px;
    line-height: 16px;
    color: #fc4528;
    margin-right: 10px;
`;

const SubmissionBoxErrorDisplay: React.FunctionComponent = (props: {
    error: string;
}) => {
    if (props.error.length > 0) {
        return <ErrorMessage>{props.error}</ErrorMessage>;
    } else {
        return <p></p>;
    }
};

export default SubmissionBoxErrorDisplay;
