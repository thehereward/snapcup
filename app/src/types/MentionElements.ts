export default interface MentionElements {
    id: String;
    display: String;
}
