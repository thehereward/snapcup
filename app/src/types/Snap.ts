export default interface Snap {
    id?: string /* Document id from FireStore */;
    to: String[] /* Recipient UIDs */;
    from: string /* Sender UID */;
    body: string /* Main Message */;
    timestamp: Date /* Date/time of Submission*/;
}
