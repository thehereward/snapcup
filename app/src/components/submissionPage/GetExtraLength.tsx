import MentionElements from "../../types/MentionElements";

function GetExtraLength(snappedUsers: MentionElements[]): number {
    let total = 280;
    for (var elem of snappedUsers) {
        total = total + elem.id.length + 5;
    }
    return total;
}

export default GetExtraLength;
