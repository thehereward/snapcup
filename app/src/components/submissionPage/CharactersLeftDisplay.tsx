import MentionElements from "../../types/MentionElements";
import GetExtraLength from "./GetExtraLength";
import React from "react";
import styled from "styled-components";

import { HelperText } from "./SnapSubmissionStyles";

const CharactersLeft = styled(HelperText)`
    text-align: right;
    color: #faf8f8;
`;

/* Calculate the amount of characters left for you to use*/
export const GetCharacterCountRemaining = (
    message: string,
    snappedUsers: MentionElements[]
) => {
    return GetExtraLength(snappedUsers) - message.length;
};

const CharactersLeftDisplay: React.FunctionComponent = (props: {
    snappedUsers: MentionElements[];
    message: string;
}) => {
    return (
        <CharactersLeft>
            There are{" "}
            {GetCharacterCountRemaining(props.message, props.snappedUsers)}{" "}
            characters remaining
        </CharactersLeft>
    );
};

export default CharactersLeftDisplay;
