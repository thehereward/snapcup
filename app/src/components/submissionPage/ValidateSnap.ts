import MentionElements from "~types/MentionElements";
import Snap from "../../types/Snap";
import { GetCharacterCountRemaining } from "./CharactersLeftDisplay";

/* Validates a snap */
const ValidateSnap = (snap: Snap, snappedUsers: MentionElements[]): Boolean => {
    /* Conditions for Valid Snap:
        1. Message Length (without metadata) < 280 
        2. From field must not be empty
        3. To field must not be undefined
        4. Timestamp field must not be empty
    */
    const condition1: boolean =
        GetCharacterCountRemaining(snap.body, snappedUsers) >= 0;
    const condition2: boolean = snap.from && snap.from.length > 0;
    const condition3: boolean = !(snap.to === undefined);
    const condition4: boolean = snap.timestamp != null;
    return condition1 && condition2 && condition3 && condition4;
};

export default ValidateSnap;
