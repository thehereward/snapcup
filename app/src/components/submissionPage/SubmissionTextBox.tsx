import React from "react";
import { useState } from "react";
import { MentionsInput, Mention } from "react-mentions";
import MentionElements from "../../types/MentionElements";
import GetExtraLength from "./GetExtraLength";
import CharactersLeftDisplay from "./CharactersLeftDisplay";
import OnSubmitMessageDisplay from "./OnSubmitMessageDisplay";
import SubmissionBoxErrorDisplay from "./SubmissionBoxErrorDisplay";
import ValidateSnap from "./ValidateSnap";
import { getCurrentUserUid } from "../../firebase/users/UserService";
import Snap from "../../types/Snap";
import { submitSnap } from "../../firebase/snaps/SnapService";
import {
    ElleImg,
    SnapSubmissionColumnDiv,
    SnapCupTextArea,
    SnapCupText,
    SnapItButton,
    LabelText,
    TaggedTeamMembers,
    HelperText,
} from "./SnapSubmissionStyles";

export interface Props {
    snappables: MentionElements[];
    user: string;
}

const SubmissionTextBox: React.FunctionComponent = (props: Props) => {
    /* Containing body of the snap */
    const [message, setMessage] = useState<string>("");
    const [confirmation, setConfirmation] = useState<Boolean>(false);
    const [error, setError] = useState<string>("");
    const [snappedUsers, setSnappedUsers] = useState<MentionElements[]>([]);

    function handleSubmit(event) {
        const uid = getCurrentUserUid();
        event.preventDefault();
        const ids = snappedUsers.map((u: MentionElements) => u.id);
        const resultingSnap: Snap = {
            to: ids,
            from: uid,
            body: message,
            timestamp: new Date(),
        };
        if (ValidateSnap(resultingSnap, snappedUsers)) {
            try {
                submitSnap(resultingSnap);
                setConfirmation(true);
                setMessage("");
                setSnappedUsers([]);
            } catch (error) {
                setError(error.toString());
            }
        } else {
            setError("Your snap is invalid.");
            setConfirmation(false);
        }
    }

    /* Updates the value in the webhook "message" */
    function handleMessageTextChanged(
        event,
        newValue,
        newPlainTextValue,
        mentions
    ) {
        setMessage(event.target.value);
        setConfirmation(false);
        setError("");
        setSnappedUsers(mentions);
    }

    return (
        <div className="container-sm">
            <SnapSubmissionColumnDiv className="row justify-content-md-center">
                <div className="d-none d-sm-block col col-lg-5 ">
                    <SnapCupText>
                        Add a Snap to the current SnapCup.
                    </SnapCupText>
                    <ElleImg className="w-100" />
                </div>
                <div className="col col-lg-5">
                    <form>
                        <div>
                            <LabelText>Tag Team Members</LabelText>
                            <TaggedTeamMembers
                                type="text"
                                className="form-control"
                            />
                            <HelperText>
                                Enter their email address or use the @ symbol
                            </HelperText>
                        </div>
                        <div className="form-group">
                            <LabelText>Message:</LabelText>
                            <SnapCupTextArea
                                className="form-control"
                                value={message}
                                onChange={handleMessageTextChanged}
                                maxLength={GetExtraLength(snappedUsers)}
                                rows={5}
                                placeholder="You can tag users using @."
                            >
                                <Mention
                                    style={{
                                        backgroundColor: "#daf4fa",
                                        zIndex: 0,
                                    }}
                                    trigger="@"
                                    data={props.snappables}
                                    rows={5}
                                />
                            </SnapCupTextArea>
                            {confirmation ? (
                                <OnSubmitMessageDisplay
                                    confirmation={confirmation}
                                />
                            ) : (
                                <CharactersLeftDisplay
                                    snappedUsers={snappedUsers}
                                    message={message}
                                />
                            )}
                            {error ? (
                                <SubmissionBoxErrorDisplay error={error} />
                            ) : null}
                        </div>
                        <SnapItButton
                            type="submit"
                            className="btn btn-primary"
                            onClick={handleSubmit}
                        >
                            Snap it
                        </SnapItButton>
                    </form>
                </div>
            </SnapSubmissionColumnDiv>
        </div>
    );
};

export default SubmissionTextBox;
