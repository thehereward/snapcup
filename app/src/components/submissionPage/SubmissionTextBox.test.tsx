// SubmissionTextBox.test.tsx
import React from "react";
import renderer from "react-test-renderer";
import { render as librender } from "@testing-library/react";
import { screen } from "@testing-library/react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import MentionElements from "../../types/MentionElements";
import SubmissionTextBox, { Props } from "./SubmissionTextBox";
import GetExtraLength from "./GetExtraLength";
import CharactersLeftDisplay, {
    GetCharacterCountRemaining,
} from "./CharactersLeftDisplay";
import OnSubmitMessageDisplay from "./OnSubmitMessageDisplay";
import Snap from "../../firebase/snap/Snap";
import ValidateSnap from "./ValidateSnap";
import SubmissionBoxErrorDisplay from "./SubmissionBoxErrorDisplay";

let container = null;
beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
});

it("renders without crashing", () => {
    act(() => {
        render(
            <SubmissionTextBox
                snappables={[{ id: "test-id", display: "test-display" }]}
                user={"test-user"}
            />,
            container
        );
    });
    expect(container.children.length).toBeGreaterThan(0);
});

describe("CharactersLeftDisplay", () => {
    test("characters remaining string renders", () => {
        const elem1: MentionElements = { id: "idlength10", display: "Jim" };
        const elem2: MentionElements = {
            id: "anna@softwire.com",
            display: "Anna",
        };
        const msg2: String =
            "Hi there @[Anna](anna@softwire.com) and @[Jim](idlength10), you have cool voices!";
        const component = renderer.create(
            <CharactersLeftDisplay
                snappedUsers={[elem1, elem2]}
                message={msg2}
            />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
    test("characters remaining string renders", () => {
        const elem1: MentionElements = { id: "idlength10", display: "Jim" };
        const elem2: MentionElements = {
            id: "anna@softwire.com",
            display: "Anna",
        };
        const msg1: String = "";
        const component = renderer.create(
            <CharactersLeftDisplay snappedUsers={[]} message={msg1} />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe("OnSubmitMessageDisplay", () => {
    test("Submit message appears if confirmation is true", () => {
        librender(<OnSubmitMessageDisplay confirmation={true} />);
        screen.getByText("Snap submitted!");
    });
    test("Submit message is empty if confirmation is false", () => {
        librender(<OnSubmitMessageDisplay confirmation={false} />);
        const text = screen.queryByText("Snap submitted!");
        expect(text).toBe(null);
    });
});

describe("SubmissionBoxErrorDisplay", () => {
    test("test for appearance of error message on error", () => {
        librender(<SubmissionBoxErrorDisplay error={"error"} />);
        screen.getByText("error");
        const text = screen.queryAllByText(/./);
        expect(text.length).toBeGreaterThan(0);
    });
    test("test for no error message if no error", () => {
        librender(<SubmissionBoxErrorDisplay error={""} />);
        const text = screen.queryAllByText(/./);
        expect(text.length).toBe(0);
    });
});

/*  */
describe("<SubmissionTextBox />", () => {
    test("test for successful submission of a legitimate snap", () => {
        /*  Ensure that snaps are legitimate before firebase submission */
        const elem1: MentionElements = {
            id: "RandomUNiqueID1",
            display: "Jimathy",
        };
        const emptyMentions: MentionElements[] = [];
        const someMentions: MentionElements[] = [elem1];
        const message1: String = "Hello mate, how are you?";
        const message2: String =
            "Hello @(Jimathy)[RandomUNiqueID1], you are cool";
        const validSnap1: Snap = {
            body: message1,
            to: [],
            from: "Timathy",
            timestamp: new Date(),
        };
        const validSnap2: Snap = {
            body: message2,
            to: ["RandomUNiqueID1"],
            from: "Timathy",
            timestamp: new Date(),
        };
        //expect(ValidateSnap(validSnap1, emptyMentions)).toBe(true)
        expect(ValidateSnap(validSnap2, someMentions)).toBe(true);
    });
    test("test for character count changes on type and appearance of characters left", () => {
        /*  As you type in the character count changes
            "Helloooo" -> chacter coutner = 272
            "Hello @user" -> character counter = 269
        */

        const elem1: MentionElements = {
            id: "RandomUNiqueID1",
            display: "Jimathy",
        };
        const emptyMentions: MentionElements[] = [];
        const someMentions: MentionElements[] = [elem1];
        const message1: String = "Hello mate, how are you?";
        const message2: String =
            "Hello @(Jimathy)[RandomUNiqueID1], you are cool";
        expect(GetCharacterCountRemaining(message1, emptyMentions)).toBe(
            280 - message1.length
        );
        expect(GetCharacterCountRemaining(message2, someMentions)).toBe(
            280 - message2.length + "@()[]".length + elem1.id.length
        );
    });

    test("test for character limit changes when tagged users changes", () => {
        /*  React Mentions includes metadata inside the value field making a tag like 
            @tom  -> @(tom)[uniqueID]
            This function should adjust the character count to ignore metadata.
        */
        const emptyMentions: MentionElements[] = [];
        const elem1: MentionElements = {
            id: "RandomUNiqueID1",
            display: "Jimathy",
        };
        const elem2: MentionElements = {
            id: "RandomUNiqueID2",
            display: "Jonathy",
        };
        const elem3: MentionElements = {
            id: "RandomUNiqueID2",
            display: "Jenathy",
        };
        const someMentions: MentionElements[] = [elem1, elem2, elem3];
        expect(GetExtraLength(emptyMentions)).toBe(280);
        expect(GetExtraLength(someMentions)).toBe(
            280 + 3 * (elem1.id.length + 5)
        );
    });
});
