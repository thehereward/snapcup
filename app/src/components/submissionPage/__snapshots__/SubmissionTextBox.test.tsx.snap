// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharactersLeftDisplay characters remaining string renders 1`] = `
<p>
  There are 
  236
   characters remaining
</p>
`;

exports[`CharactersLeftDisplay characters remaining string renders 2`] = `
<p>
  There are 
  280
   characters remaining
</p>
`;
